// alchemy-sdk/dist/cjs/api/utils.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var hash = require('@ethersproject/hash');
var bytes = require('@ethersproject/bytes');
var units = require('@ethersproject/units');



Object.defineProperty(exports, 'dnsEncode', {
	enumerable: true,
	get: function () { return hash.dnsEncode; }
});
Object.defineProperty(exports, 'hashMessage', {
	enumerable: true,
	get: function () { return hash.hashMessage; }
});
Object.defineProperty(exports, 'id', {
	enumerable: true,
	get: function () { return hash.id; }
});
Object.defineProperty(exports, 'isValidName', {
	enumerable: true,
	get: function () { return hash.isValidName; }
});
Object.defineProperty(exports, 'namehash', {
	enumerable: true,
	get: function () { return hash.namehash; }
});
Object.defineProperty(exports, 'arrayify', {
	enumerable: true,
	get: function () { return bytes.arrayify; }
});
Object.defineProperty(exports, 'concat', {
	enumerable: true,
	get: function () { return bytes.concat; }
});
Object.defineProperty(exports, 'hexConcat', {
	enumerable: true,
	get: function () { return bytes.hexConcat; }
});
Object.defineProperty(exports, 'hexDataLength', {
	enumerable: true,
	get: function () { return bytes.hexDataLength; }
});
Object.defineProperty(exports, 'hexDataSlice', {
	enumerable: true,
	get: function () { return bytes.hexDataSlice; }
});
Object.defineProperty(exports, 'hexStripZeros', {
	enumerable: true,
	get: function () { return bytes.hexStripZeros; }
});
Object.defineProperty(exports, 'hexValue', {
	enumerable: true,
	get: function () { return bytes.hexValue; }
});
Object.defineProperty(exports, 'hexZeroPad', {
	enumerable: true,
	get: function () { return bytes.hexZeroPad; }
});
Object.defineProperty(exports, 'hexlify', {
	enumerable: true,
	get: function () { return bytes.hexlify; }
});
Object.defineProperty(exports, 'isBytes', {
	enumerable: true,
	get: function () { return bytes.isBytes; }
});
Object.defineProperty(exports, 'isBytesLike', {
	enumerable: true,
	get: function () { return bytes.isBytesLike; }
});
Object.defineProperty(exports, 'isHexString', {
	enumerable: true,
	get: function () { return bytes.isHexString; }
});
Object.defineProperty(exports, 'joinSignature', {
	enumerable: true,
	get: function () { return bytes.joinSignature; }
});
Object.defineProperty(exports, 'splitSignature', {
	enumerable: true,
	get: function () { return bytes.splitSignature; }
});
Object.defineProperty(exports, 'stripZeros', {
	enumerable: true,
	get: function () { return bytes.stripZeros; }
});
Object.defineProperty(exports, 'zeroPad', {
	enumerable: true,
	get: function () { return bytes.zeroPad; }
});
Object.defineProperty(exports, 'formatEther', {
	enumerable: true,
	get: function () { return units.formatEther; }
});
Object.defineProperty(exports, 'formatUnits', {
	enumerable: true,
	get: function () { return units.formatUnits; }
});
Object.defineProperty(exports, 'parseEther', {
	enumerable: true,
	get: function () { return units.parseEther; }
});
Object.defineProperty(exports, 'parseUnits', {
	enumerable: true,
	get: function () { return units.parseUnits; }
});
